cfme/common/vm_console.py

```diff
@@ -0,0 +1,179 @@
+# -*- coding: utf-8 -*-
+"""
+Module containing classes with common behaviour for consoles of both VMs and Instances of all types.
+"""
+
+import base64
+import re
+import tempfile
+
+from PIL import Image, ImageFilter
+from pytesseract import image_to_string
+from utils.log import logger
+from utils.pretty import Pretty
+from wait_for import wait_for, TimedOutError
+
+from cfme.cloud.provider.openstack import OpenStackProvider
+
+banner_id = 'noVNC_status'  # specific for RHOS Provider
+canvas_id = 'noVNC_canvas'
+ctrl_alt_del_id = 'sendCtrlAltDelButton'  # specific for RHOS Provider
+connection_status_xpath = '//*[@id="connection-status"]'  # specific for the VMware/RHEV Provider
+canvas_xapth = '//*[@id="remote-console"]/canvas'  # specific to VMware/RHEV Provider
+ctrl_alt_del_xpath = '//*[@id="ctrlaltdel"]'  # specific to VMware/RHEV Provider
+
+
+class VMConsole(Pretty):
+    """Class to manage the VM Console.   Presently, only support HTML5 Console."""
+
+    pretty_attrs = ['appliance_handle', 'browser', 'console_handle', 'name']
+
+    def __init__(self, name, selenium, console_handle, appliance_handle, provider):
+        self.name = name
+        self.selenium = selenium
+        self.console_handle = console_handle
+        self.appliance_handle = appliance_handle
+        self.provider = provider
+
+    ###
+    # Methods
+    #
+    def _get_openstack_banner(self):
+        text = self.selenium.find_element_by_id(banner_id).text
+        return text
+
+    def _get_vmware_rhev_connection_status_text(self):
+        text = self.selenium.find_element_by_xpath(connection_status_xpath).text
+        return text
+
+    def get_banner(self):
+        """Get the text of the banner above the console screen."""
+        self.switch_to_console()
+        if self.provider.one_of(OpenStackProvider):
+            text = self._get_openstack_banner()
+        else:
+            text = self._get_vmware_rhev_connection_status_text()
+        logger.info('Read following text from console banner: {}'.format(text))
+        self.switch_to_appliance()
+        return text
+
+    def _get_openstack_canvas(self):
+        return self.selenium.find_element_by_id(canvas_id)
+
+    def _get_vmware_rhev_canvas(self):
+        return self.selenium.find_element_by_xpath(canvas_xapth)
+
+    def get_screen(self):
+        """
+        Retrieve the bit map from the canvas widget that represents the console screen.
+
+        Returns it as a binary string.
+
+        Implementation:
+        The canvas tag has a method toDataURL() which one can use in javascript to
+        obtain the canvas image  base64 encoded.   Examples of how to do this can be
+        seen here:
+
+            https://qxf2.com/blog/selenium-html5-canvas-verify-what-was-drawn/
+            https://stackoverflow.com/questions/38316402/how-to-save-a-canvas-as-png-in-selenium
+        """
+        self.switch_to_console()
+
+        # Get the canvas element
+        canvas = self.get_canvas()
```

Comment:
These tests are entirely dependent on the fix set of Templates, PhotonOS for VMware and Cirros for OpenStack. So with those we won't face screensaver situation. For GUI based VM Consoles, I will have to find some other way to deal with it. My primary goal was being able to interact with some VM console, so I chose to go with the command line based first. It was collective decision of myself and @jamesooden . Hope it makes sense.